Allow to format signatures in docstrings

krassowski opened the pull request:
This is a proposal for #630. Initially this PR proposed adding `pylsp_signatures_to_markdown` hook, but currently it modifies the `format_docstring` directly without use of the hook.

Using a list → str approach allows the hook to modify how alternative signatures are displayed.

<s>There are no user-facing changes out of the box, however users can now install [`pylsp-ruff-signature`](https://github.com/krassowski/pylsp-ruff-signature)</s>

Users see much more readable function signatures in hover and in completion documentation, for example:

| Without `pylsp-ruff-signature` | With `pylsp-ruff-signature` |
|--|--|
| ![image](https://github.com/user-attachments/assets/5c53e95c-866e-4f12-9bd1-2d67059aa43a) | ![image](https://github.com/user-attachments/assets/41d0ffae-13a1-472b-9e6f-2b7e2bd76b9f) |

krassowski committed:
Allow to format signatures in docstrings

krassowski committed:
Add a test

krassowski commented:
Any thoughts @ccordoba12 ? It would be amazing to get it in to allow downstreams to customize.

krassowski committed:
Merge branch 'develop' into docstring-hook

krassowski commented:
A gentle ping :)

krassowski commented on pylsp/plugins/hover.py:
```diff
@@ -9,7 +9,7 @@
 
 @hookimpl
-def pylsp_hover(config, document, position):
+def pylsp_hover(config, document, position, signatures_to_markdown=None):
```
I tripped myself on https://github.com/pytest-dev/pluggy/issues/442 again.

krassowski committed:
Simplify hook usage, do not use default args

as these are not supported by pluggy

krassowski committed:
Fix tests

krassowski committed:
One more test

ccordoba12 commented:
> Any thoughts @ccordoba12 ? It would be amazing to get it in to allow downstreams to customize.

Sorry for the delay in replying @krassowski. For now I only have one question for you: instead of using a hook for this (`pylsp_signatures_to_markdown`), why not adding the functionality to format signatures to markdown to the server itself?

The problem with your current approach is that only users that know that they need to install `pylsp-ruff-signature` (or a similar plugin), will be able to get better signatures. That would be ok for JupyterLab-LSP users because you could add that plugin as a new dependency. But it wouldn't be easily discoverable for users from other editors/IDEs.

I know that would add a new dependency in either Ruff or Black, but I'm ok with that. I'd prefer Black because it's pure Python, but we could look for Ruff being on PATH and use it first if found. Another option would be to create a new package (similar to `docstring_to_markdown`) and use it for this purpose. Then, Black/Ruff would be its dependency.

krassowski commented:
> instead of using a hook for this (`pylsp_signatures_to_markdown`), why not adding the functionality to format signatures to markdown to the server itself?

I think there are two cons:
- it adds a new dependency (say `ruff` or `black`) - you addressed this point
- users may want to configure it (e.g. max line length, how to present typing, etc)

If you are ok with having this in `pylsp` and including the config options then I think would be fine.

I think we could even get away with having `ruff`/`black` as optional dependencies by introducing this plugin as disabled by default (and sniffing if `ruff`/`black` are installed).

ccordoba12 commented:
> users may want to configure it (e.g. max line length, how to present typing, etc)

Right, I didn't thought about it.

> If you are ok with having this in pylsp and including the config options then I think would be fine.

Sure, no problem about it if that'd benefit more users.

> I think we could even get away with having ruff/black as optional dependencies by introducing this plugin as disabled by default (and sniffing if ruff/black are installed).

That'd be inconsistent to users because with no ruff/black they'd get different results than without them and it'd be hard for them to understand why. So, I'd say let's go with black and leave ruff as optional. In that case, the UX would be practically indistinguishable since both give pretty much the same result.

dalthviz commented:
Just in case, gave a check to the implementation and current approach (using Jupyterlab + this PR + `pylsp-ruff-signature`) and I think things are working as expected  :+1:

Regarding the implementation approach to follow, the idea then is to include `python-ruff-signature` as a bundled plugin for `python-lsp-server` (leaving it enabled by default) + allowing it to be configurable to work with either black or ruff and other configs that could be useful (max line length, typing represetation, etc) + add `black` as a required dependency for `python-lsp-server` while `ruff` would be optional, right?

ccordoba12 commented:
> Regarding the implementation approach to follow, the idea then is to include python-ruff-signature as a bundled plugin for python-lsp-server

Nop, it's to implement the functionality of that plugin directly in `_utils/format_docstring`. Then a new hook, changes to existing hooks and a new plugin won't be necessary.

> allowing it to be configurable to work with either black or ruff and other configs that could be useful (max line length, typing represetation, etc)

Yep, although I think that's a bit overkill.

> add black as a required dependency for python-lsp-server while ruff would be optional, right?

Right.

krassowski committed:
Remove the hook, implement formatters natively

krassowski committed:
Add black as a dependency

krassowski committed:
Fixes

krassowski committed:
Fix/update config

krassowski committed:
Fix opt-out, add a test

krassowski committed:
Lint

krassowski committed:
Python 3.8 support

krassowski committed:
Merge branch 'develop' into docstring-hook

krassowski commented:
@ccordoba12 @dalthviz when you get a chance this is ready for another round of review - as per the request above I removed the hook and implemented the functionality directly, adding a `black` dependency and allowing for opting out or choosing `ruff` instead.

dalthviz commented on pyproject.toml:
```diff
@@ -19,6 +19,7 @@ dependencies = [
     "pluggy>=1.0.0",
     "python-lsp-jsonrpc>=1.1.0,<2.0.0",
     "ujson>=3.0.0",
+    "black"
```
Should `ruff` be added as an optional dependecy below?

ccordoba12 replied:
I don't think so. The functionality to format signatures will be provided by Black and it'll be very similar to the one given by Ruff.

dalthviz commented on pylsp/_utils.py:
```diff
@@ -209,8 +211,113 @@ def choose_markup_kind(client_supported_markup_kinds: List[str]):
+    if formatter.is_installed:
+        try:
+            return (
+                # TODO: replace with str.removeprefix and str.removesuffix
```
Should an issue be created to track this TODO @ccordoba12 ?

ccordoba12 replied:
No, I think we should simply bump our Python version to 3.9 in another PR so we don't need those functions here.

Could you take care of that?

krassowski replied:
I opened https://github.com/python-lsp/python-lsp-server/pull/650

dalthviz reviewed (approved):
Thank you @krassowski for the updates here! I checked locally the changes here and seems like things are working as expected :+1: 

Left a comment regarding adding `ruff` as an optional dependency and another regarding the TODO about the `removesuffix`/`removeprefix` functions and Python 3.8 but besides that this LGTM so leaving it approved.

krassowski committed:
Merge branch 'develop' into docstring-hook

krassowski committed:
Remove code for Python 3.8 compatibility
